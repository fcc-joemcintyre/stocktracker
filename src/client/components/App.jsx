/**
 * Copyright (c) Joe McIntyre, 2016-2018
 * license: MIT (https://github.com/fcc-joemcintyre/stocktracker/LICENSE.txt)
 */
import React, { Component } from 'react';
import { render } from 'react-dom';
import io from 'socket.io-client';
import { Titlebar } from './Titlebar.jsx';
import { Range } from './Range.jsx';
import { StockChart } from './StockChart.jsx';
import { Stock } from './Stock.jsx';
import { StockEntry } from './StockEntry.jsx';
import { Status } from './Status.jsx';

export default class App extends Component {
  constructor (props) {
    super (props);
    this.state = {
      errors: [],
      retrieving: [],
      data: [],
      months: 12,
      colors: ['blue', 'green', 'lightblue', 'lightgreen', 'purple', 'orange',
        'lightpurple', 'steelblue'],
      chartKey: 1,
    };
    this.addStock = this.addStock.bind (this);
    this.server = this.server.bind (this);
    this.removeStock = this.removeStock.bind (this);
    this.update = this.update.bind (this);
    this.onRangeChanged = this.onRangeChanged.bind (this);
  }

  /**
   * On component mount, establish socket connection with monitoring server
   * @return {void}
   */
  componentDidMount () {
    const socket = io.connect ();
    socket.on ('update', this.update);
  }

  /**
   * Change range of months to display
   * @param {number} months Number of months in range
   * @return {void}
   */
  onRangeChanged (months) {
    this.setState ({ months });
  }

  /**
   * Add a stock to set of monitored stocks. Initially add it to the list
   * of requests in process (retrieving). The update function will change
   * the state from retrieving to monitored.
   * @param {string} symbol Stock trading symbol
   * @return {void}
   */
  addStock (symbol) {
    this.server ('PUT', symbol);
    this.state.retrieving.push (symbol);
    this.forceUpdate ();
  }

  /**
   * Initiate interaction with server. Data will be received through
   * the update function from broadcasts from the monitoring server.
   * @param {string} action PUT or DELETE
   * @param {string} symbol Stock trading symbol
   * @return {void}
   */
  server (action, symbol) {
    const req = new XMLHttpRequest ();
    req.onreadystatechange = (() => {
      if (req.readyState === 4) {
        if (req.status !== 200) {
          this.setState (prev => ({ errors: prev.errors.concat (`Error processing ${symbol}`) }));
        }
      }
    });
    req.open (action, `api/stocks/${symbol}`, true);
    req.send ();
  }

  /**
   * Remove stock from interest (remove from retrieving or monitored set).
   * @param {string} symbol Stock trading symbol
   * @return {void}
   */
  removeStock (symbol) {
    this.server ('DELETE', symbol);
  }

  /**
   * On receipt of monitor broadcast, update retrieving list and then
   * set new data.
   * @param {string} input Data from monitoring server
   * @return {void}
   */
  update (input) {
    const data = JSON.parse (input);
    this.setState ((prev) => {
      // remove retrieving items as received
      const retrieving = prev.retrieving.filter ((a) => {
        for (const item of data) {
          if (item.symbol === a) {
            return false;
          }
        }
        return true;
      });

      // divide error and completed data items
      const errors = [];
      const filtered = data.filter ((item) => {
        if (item.status !== 0) {
          const text = (item.status >= 500) ? 'Stock quote service unavailable' :
            (item.status === 404) ? 'Stock symbol not found' :
              (item.status === 429) ? 'Request limit reached' : 'Unknown';
          errors.push (`[Stock:${item.symbol} Error:${item.status} (${text})]`);
          return false;
        }
        return true;
      });
      return ({
        retrieving,
        data: filtered,
        errors,
        chartKey: prev.chartKey + 1,
      });
    });
  }

  render () {
    const stocks = [];
    let color = 0;
    for (const item of this.state.data) {
      // assign color to item
      item.color = this.state.colors[color];
      stocks.push (
        <Stock
          key={item.symbol}
          symbol={item.symbol}
          name={item.name}
          color={item.color}
          removeStock={this.removeStock}
        />
      );
      color += 1;
    }

    // if not all tracking slots (8) are taken, show entry component
    const stockEntry = (stocks.length < 8) ? <StockEntry addStock={this.addStock} /> : null;

    return (
      <div>
        <Titlebar />
        <Range months={this.state.months} onRangeChanged={this.onRangeChanged} />
        <StockChart
          key={this.state.chartKey}
          className='stockChart'
          width={900}
          height={500}
          months={this.state.months}
          stocks={this.state.data}
        />
        <div className='stockArea'>
          {stocks}
          {stockEntry}
        </div>
        <Status requests={this.state.retrieving} errors={this.state.errors} />
      </div>
    );
  }
}

render (<App />, document.getElementById ('app'));
