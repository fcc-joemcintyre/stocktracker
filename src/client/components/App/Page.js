import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Titlebar } from '../Titlebar';
import { Range } from '../Range';
import { StockChart } from '../StockChart';
import { Stock } from '../Stock';
import { StockEntry } from '../StockEntry';
import { Status } from '../Status';

export const Page = ({ data, months, retrieving, errors, chartKey, onRangeChanged, onAddStock, onRemoveStock }) => {
  const colors = ['blue', 'green', 'lightblue', 'lightgreen', 'purple', 'orange', 'lightpurple', 'steelblue'];
  const stockData = data.map ((item, index) => Object.assign ({}, item, { color: colors[index] }));
  const stocks = data.map ((item, index) => (
    <Stock
      key={item.symbol}
      symbol={item.symbol}
      name={item.name}
      color={colors[index]}
      onRemoveStock={onRemoveStock}
    />
  ));

  // if not all tracking slots (8) are taken, show entry component
  const stockEntry = (stocks.length < 8) ? <StockEntry onAddStock={onAddStock} /> : null;

  return (
    <Fragment>
      <Titlebar />
      <Range months={months} onRangeChanged={onRangeChanged} />
      <StockChart
        key={chartKey}
        className='stockChart'
        width={900}
        height={500}
        months={months}
        stocks={stockData}
      />
      <div className='stockArea'>
        {stocks}
        {stockEntry}
      </div>
      <Status requests={retrieving} errors={errors} />
    </Fragment>
  );
};

Page.propTypes = {
  data: PropTypes.arrayOf (PropTypes.shape ({})).isRequired,
  months: PropTypes.number.isRequired,
  retrieving: PropTypes.arrayOf (PropTypes.string).isRequired,
  errors: PropTypes.arrayOf (PropTypes.string).isRequired,
  chartKey: PropTypes.number.isRequired,
  onRangeChanged: PropTypes.func.isRequired,
  onAddStock: PropTypes.func.isRequired,
  onRemoveStock: PropTypes.func.isRequired,
};
