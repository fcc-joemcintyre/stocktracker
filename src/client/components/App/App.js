import React, { Component } from 'react';
import io from 'socket.io-client';
import { Page } from './Page';

export class App extends Component {
  constructor (props) {
    super (props);
    this.state = {
      errors: [],
      retrieving: [],
      data: [],
      months: 12,
      chartKey: 1,
    };
    this.onRangeChanged = this.onRangeChanged.bind (this);
    this.onAddStock = this.onAddStock.bind (this);
    this.onRemoveStock = this.onRemoveStock.bind (this);
    this.server = this.server.bind (this);
    this.update = this.update.bind (this);
  }

  /**
   * On component mount, establish socket connection with monitoring server
   * @return {void}
   */
  componentDidMount () {
    const socket = io.connect ();
    socket.on ('update', this.update);
  }

  /**
   * Change range of months to display
   * @param {number} months Number of months in range
   * @return {void}
   */
  onRangeChanged (months) {
    this.setState ({ months });
  }

  /**
   * Add a stock to set of monitored stocks. Initially add it to the list
   * of requests in process (retrieving). The update function will change
   * the state from retrieving to monitored.
   * @param {string} symbol Stock trading symbol
   * @return {void}
   */
  onAddStock (symbol) {
    this.server ('PUT', symbol);
    this.state.retrieving.push (symbol);
    this.forceUpdate ();
  }

  /**
   * Remove stock from interest (remove from retrieving or monitored set).
   * @param {string} symbol Stock trading symbol
   * @return {void}
   */
  onRemoveStock (symbol) {
    this.server ('DELETE', symbol);
  }

  /**
   * Initiate interaction with server. Data will be received through
   * the update function from broadcasts from the monitoring server.
   * @param {string} action PUT or DELETE
   * @param {string} symbol Stock trading symbol
   * @return {void}
   */
  server (action, symbol) {
    const req = new XMLHttpRequest ();
    req.onreadystatechange = (() => {
      if (req.readyState === 4) {
        if (req.status !== 200) {
          this.setState (prev => ({ errors: prev.errors.concat (`Error processing ${symbol}`) }));
        }
      }
    });
    req.open (action, `api/stocks/${symbol}`, true);
    req.send ();
  }

  /**
   * On receipt of monitor broadcast, update retrieving list and then
   * set new data.
   * @param {string} input Data from monitoring server
   * @return {void}
   */
  update (input) {
    const data = JSON.parse (input);
    this.setState ((prev) => {
      // remove retrieving items as received
      const retrieving = prev.retrieving.filter ((a) => {
        for (const item of data) {
          if (item.symbol === a) {
            return false;
          }
        }
        return true;
      });

      // divide error and completed data items
      const errors = [];
      const filtered = data.filter ((item) => {
        if (item.status !== 0) {
          const text = (item.status >= 500) ? 'Stock quote service unavailable' :
            (item.status === 404) ? 'Stock symbol not found' :
              (item.status === 429) ? 'Request limit reached' : 'Unknown';
          errors.push (`[Stock:${item.symbol} Error:${item.status} (${text})]`);
          return false;
        }
        return true;
      });
      // convert dates in data itmes to Date objects
      const translated = filtered.map (
        a => Object.assign ({}, a, { data: a.data.map (
          b => Object.assign ({}, b, { 0: new Date (b[0]) })
        ) })
      );
      return ({
        retrieving,
        data: translated,
        errors,
        chartKey: prev.chartKey + 1,
      });
    });
  }

  render () {
    return (
      <Page
        data={this.state.data}
        months={this.state.months}
        retrieving={this.state.retrieving}
        errors={this.state.errors}
        chartKey={this.state.chartKey}
        onRangeChanged={this.onRangeChanged}
        onAddStock={this.onAddStock}
        onRemoveStock={this.onRemoveStock}
      />
    );
  }
}
