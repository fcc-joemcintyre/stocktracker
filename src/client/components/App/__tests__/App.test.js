/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import * as mockSocket from 'mock-socket';
import { mount } from 'enzyme';
import { App } from '../App';

test ('App starting state', () => {
  const component = renderer.create (
    <App />
  );
  const tree = component.toJSON ();
  expect (tree).toMatchSnapshot ();
});

jest.mock ('socket.io-client', () => mockSocket.SocketIO);

test ('App starting state', () => {
  const mockServer = new mockSocket.Server ('ws://localhost:3999');
  mockServer.on ('connection', () => {
    mockServer.emit ('connect');
    mockServer.on ('subscribe', (data, fn) => fn ());
  });

  mount (<App />);
});
