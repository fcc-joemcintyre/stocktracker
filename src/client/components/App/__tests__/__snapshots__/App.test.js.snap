// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App starting state 1`] = `
Array [
  <div
    className="titleArea"
  >
    <div
      className="titleText"
    >
      Stock Tracker
    </div>
  </div>,
  <div
    className="rangeBox"
  >
    <span
      className=""
      onClick={[Function]}
    >
      1m
    </span>
    <span
      className=""
      onClick={[Function]}
    >
      3m
    </span>
    <span
      className=""
      onClick={[Function]}
    >
      6m
    </span>
    <span
      className="selected"
      onClick={[Function]}
    >
      1y
    </span>
    <span
      className=""
      onClick={[Function]}
    >
      3y
    </span>
  </div>,
  <svg
    height={500}
    width={900}
  />,
  <div
    className="stockArea"
  >
    <form
      className="entryBox"
      onSubmit={[Function]}
    >
      <label
        htmlFor="symbolText"
      >
        Symbol:
      </label>
      <input
        className="entryText"
        id="symbolText"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className="entryButton"
        disabled={true}
        type="submit"
      >
        Track
      </button>
    </form>
  </div>,
  <div
    className="statusBox"
  >
    <div>
      Pending requests: 
      none
    </div>
    <div>
      Errors: 
      none
    </div>
  </div>,
]
`;
